File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head:
            return None
        mid_pointer = head
        faster_pointer = head.next          # FIX: start one ahead so first half is longer

        while faster_pointer and faster_pointer.next:
            mid_pointer = mid_pointer.next
            faster_pointer = faster_pointer.next.next   # FIX: two steps, not one

        reversed_half = None
        tmp = None

        second = mid_pointer.next           # FIX: split the list
        mid_pointer.next = None             # FIX: otherwise you build a cycle
        mid_pointer = second

        while mid_pointer:
            tmp = mid_pointer.next
            mid_pointer.next = reversed_half
            reversed_half = mid_pointer
            mid_pointer = tmp

        dummy = res = ListNode()
        while reversed_half and head:
            next_head = head.next                   # FIX: save both successors
            next_rev = reversed_half.next           #      BEFORE rewiring
            dummy.next = head
            head.next = reversed_half
            dummy = reversed_half
            head = next_head
            reversed_half = next_rev

        if reversed_half:
            dummy.next = reversed_half
        if head:
            dummy.next = head               # FIX: was dummy.next = dummy (self-loop)
        # FIX: removed `head = res.next` — rebinding a local does nothing,
        #      and the list is already reordered in place

File: Data Structures & Algorithms/add-two-numbers/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carryBit = 0;
        ListNode res(0);
        ListNode* current = &res;
        while (l1 != nullptr && l2 != nullptr){
            current->next = new ListNode((carryBit + l1->val + l2->val) % 10);
            carryBit = (carryBit + l1->val + l2->val) / 10;
            l1 = l1->next;
            l2 = l2->next;
            current = current->next;
        }
        while (l1 != nullptr) {
            current->next = new ListNode((carryBit + l1->val) % 10);
            carryBit = (carryBit + l1->val) / 10;
            l1 = l1->next;
            current = current->next;
        }
        while (l2 != nullptr) {
            current->next = new ListNode((carryBit + l2->val) % 10);
            carryBit = (carryBit + l2->val) / 10;
            l2 = l2->next;
            current = current->next;
        }
        if (carryBit > 0) {
            current->next = new ListNode(carryBit);
            current = current->next;
            return res.next;
        }
        current->next = nullptr;
        return res.next;
    }
};
